### PR TITLE
CRNS-115/ChannelList

### PR DESCRIPTION
**Pull request by DanC5**
# Submit a pull request

## CLA

- [ ] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required).
- [ ] The code changes follow best practices
- [ ] Code changes are tested (add some information if not applicable)

### CODE REVIEW

File: src/components/ChannelList/ChannelList.js [hunk 1]
```diff
@@ -1,743 +1,270 @@
...
+      forceUpdate={forceUpdate}
+      hasNextPage={hasNextPage}
+      loadingChannels={status.loadingChannels}
+      loadNextPage={loadNextPage}
+      loadingNextPage={status.loadingNextPage}
+      refreshing={status.refreshing}
+      refreshList={refreshList}
+      reloadList={reloadList}
+      setActiveChannel={onSelect}
+      setFlatListRef={(ref) => {
```
Comment (DanC5) [inline src/components/ChannelList/ChannelList.js:90]: @vishalnarkhede can you check that this ref logic is correct? i'm not totally sure

File: src/components/MessageInput/MessageInput.js [hunk 2]
```diff
@@ -446,7 +446,7 @@ class MessageInput extends PureComponent {
     this.sending = true;
 
     const { text } = this.state;
-    await this.setState({ text: '' }, () => this.inputBox.clear());
+    await this.setState({ text: '' }, () => this.inputBox?.clear());
```
Comment (virdesai) [inline src/components/MessageInput/MessageInput.js:449]: this is gonna be fun with my PR

File: src/components/ChannelList/hooks/listeners/useNewMessageNotification.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+import { ChatContext } from '../../../../context';
+import { getChannel } from '../../utils';
+
+export const useNewMessageNotification = ({ onMessageNew, setChannels }) => {
+  const { client } = useContext(ChatContext);
+
+  useEffect(() => {
+    const handleEvent = async (e) => {
+      if (onMessageNew && typeof onMessageNew === 'function') {
```
Comment (virdesai) [inline src/components/ChannelList/hooks/listeners/useNewMessageNotification.js:12]: do you actually need the first part as if it's null or undefined or 'falsy' it wouldn't pass the typeof check
```suggestion
      if (typeof onMessageNew === 'function') {
```

File: src/components/ChannelList/hooks/listeners/useNewMessageNotification.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        onMessageNew(setChannels, e);
+      } else {
+        const channel = await getChannel(
+          client,
+          e.channel?.type,
+          e.channel?.id,
```
Comment (virdesai) [inline src/components/ChannelList/hooks/listeners/useNewMessageNotification.js:18]: ternaries mess with docs at all?
```suggestion
          e.channel?.type,
          e.channel?.id,
```

File: src/components/ChannelList/hooks/listeners/useNewMessage.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import { ChatContext } from '../../../../context';
+import { moveChannelUp } from '../../utils';
+
+export const useNewMessage = ({ lockChannelOrder, setChannels }) => {
+  const { client } = useContext(ChatContext);
+
+  useEffect(() => {
+    const handleEvent = (e) => {
+      setChannels((channels) => {
+        if (!lockChannelOrder) return moveChannelUp(e.cid, channels);
```
Comment (virdesai) [inline src/components/ChannelList/hooks/listeners/useNewMessage.js:12]: personally still don't love this single line no brackets if statement style. shortens the code but also I think it negatively effects readability a super small bit

File: src/components/ChannelList/hooks/listeners/useChannelUpdated.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+          if (index >= 0 && e.channel) {
+            channels[index].data = e.channel;
+          }
+          return [...channels];
+        });
+      }
+    };
+
+    client.on('channel.updated', handleEvent);
+    return () => client.off('channel.updated', handleEvent);
```
Comment (virdesai) [inline src/components/ChannelList/hooks/listeners/useChannelUpdated.js:26]: something we should bring up to the team for future changes to the js lib as I've seen it in some other libraries. having the return of the `client.on` function be the `'off'` function. so then all we'd need to do here would be `const channelUpdatedEventHandler = client.on('channel.updated', handleEvent);` and then `return channelUpdatedEventHandler;`

File: src/components/ChannelList/hooks/listeners/useChannelUpdated.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import { useContext, useEffect } from 'react';
+
+import { ChatContext } from '../../../../context';
+
+export const useChannelUpdated = ({ onChannelUpdated, setChannels }) => {
+  const { client } = useContext(ChatContext);
+
+  useEffect(() => {
+    const handleEvent = (e) => {
+      if (onChannelUpdated && typeof onChannelUpdated === 'function') {
```
Comment (virdesai) [inline src/components/ChannelList/hooks/listeners/useChannelUpdated.js:10]: made this comment in few places
```suggestion
      if (typeof onChannelUpdated === 'function') {
```

File: src/components/ChannelList/hooks/listeners/useChannelTruncated.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+export const useChannelTruncated = ({
+  onChannelTruncated,
+  setChannels,
+  setForceUpdate,
+}) => {
+  const { client } = useContext(ChatContext);
+
+  useEffect(() => {
+    const handleEvent = (e) => {
+      if (onChannelTruncated && typeof onChannelTruncated === 'function') {
```
Comment (virdesai) [inline src/components/ChannelList/hooks/listeners/useChannelTruncated.js:14]: made this comment in few places
```suggestion
      if (typeof onChannelTruncated === 'function') {
```

File: src/components/ChannelList/hooks/listeners/useChannelHidden.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
+import { useContext, useEffect } from 'react';
+
+import { ChatContext } from '../../../../context';
+
+export const useChannelHidden = ({ onChannelHidden, setChannels }) => {
+  const { client } = useContext(ChatContext);
+
+  useEffect(() => {
+    const handleEvent = (e) => {
+      if (onChannelHidden && typeof onChannelHidden === 'function') {
```
Comment (virdesai) [inline src/components/ChannelList/hooks/listeners/useChannelHidden.js:10]: made this comment in few places
```suggestion
      if (typeof onChannelHidden === 'function') {
```

File: src/components/ChannelList/hooks/listeners/useChannelDeleted.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
+import { useContext, useEffect } from 'react';
+
+import { ChatContext } from '../../../../context';
+
+export const useChannelDeleted = ({ onChannelDeleted, setChannels }) => {
+  const { client } = useContext(ChatContext);
+
+  useEffect(() => {
+    const handleEvent = (e) => {
+      if (onChannelDeleted && typeof onChannelDeleted === 'function') {
```
Comment (virdesai) [inline src/components/ChannelList/hooks/listeners/useChannelDeleted.js:10]: made this comment in few places
```suggestion
      if (typeof onChannelDeleted === 'function') {
```

File: src/components/ChannelList/hooks/listeners/useAddedToChannelNotification.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+export const useAddedToChannelNotification = ({
+  onAddedToChannel,
+  setChannels,
+}) => {
+  const { client } = useContext(ChatContext);
+
+  useEffect(() => {
+    const handleEvent = async (e) => {
+      if (onAddedToChannel && typeof onAddedToChannel === 'function') {
```
Comment (virdesai) [inline src/components/ChannelList/hooks/listeners/useAddedToChannelNotification.js:15]: made this comment in few places
```suggestion
      if (typeof onAddedToChannel === 'function') {
```

File: src/components/ChannelList/hooks/listeners/useChannelUpdated.js (hunk 6 above)
Comment (vishalnarkhede) [reply src/components/ChannelList/hooks/listeners/useChannelUpdated.js:26]: We already do that!! e.g., 

```js
const channelUpdatedEventHandler = client.on('channel.updated', handleEvent);

// While unmounting
channelUpdatedEventHandler.unsubscribe();
```

Comment (virdesai) [reply src/components/ChannelList/hooks/listeners/useChannelUpdated.js:26]: sweet

File: src/components/ChannelList/hooks/usePaginatedChannels.js [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+    if (queryType === 'reload') {
+      setChannels([]);
+      setLoadingChannels(true);
+    } else if (!queryType) {
+      setLoadingNextPage(true);
+    }
+
+    const newOptions = {
+      limit: options?.limit ?? MAX_QUERY_CHANNELS_LIMIT,
```
[the remark this answers is not in the document]
Comment (DanC5) [reply src/components/ChannelList/hooks/usePaginatedChannels.js:31]: the docs don't pull from any of these hooks as we pump in sample data, so all good

File: src/components/ChannelList/hooks/listeners/useRemovedFromChannelNotification.js [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+}) => {
+  const { client } = useContext(ChatContext);
+
+  useEffect(() => {
+    const handleEvent = (e) => {
+      if (onRemovedFromChannel && typeof onRemovedFromChannel === 'function') {
+        onRemovedFromChannel(setChannels, e);
+      } else {
+        setChannels((channels) => {
+          const newChannels = channels.filter((c) => c.cid !== e.channel?.cid);
```
[the remark this answers is not in the document]
Comment (vishalnarkhede) [reply src/components/ChannelList/hooks/listeners/useRemovedFromChannelNotification.js:17]: Not messing up, atleast for now.

File: src/components/ChannelList/ChannelList.js [hunk 14]
```diff
@@ -1,743 +1,270 @@
...
-      this.setState(newState, resolve);
-    });
-  };
-
-  wait = (ms) =>
-    new Promise((resolve) => {
-      setTimeout(resolve, ms);
-    });
-
-  queryChannelsRequest = async (filters, sort, options, retryCount = 1) => {
```
Comment (vishalnarkhede) [inline src/components/ChannelList/ChannelList.js:290]: Did we remove this logic? About retries

File: src/components/ChannelList/ChannelListMessenger.js [hunk 15]
```diff
@@ -1,224 +1,270 @@
...
+          renderItem={({ item: channel }) => (
+            <ChannelPreview
+              {...props}
+              channel={channel}
+              key={channel.cid}
+              Preview={Preview}
+              setActiveChannel={setActiveChannel}
+            />
+          )}
+          testID={'channel-list-messenger'}
```
Comment (virdesai) [inline src/components/ChannelList/ChannelListMessenger.js:111]:
```suggestion
          testID='channel-list-messenger'
```

File: src/components/ChannelList/ChannelListMessenger.js [hunk 16]
```diff
@@ -1,224 +1,270 @@
...
 
-  render() {
-    if (this.props.error && this.props.channels.length === 0) {
-      return this.renderLoadingError();
-    } else if (this.props.loadingChannels) {
-      return this.renderLoading();
-    } else {
-      return this.renderChannels();
-    }
+  if (error && !refreshing && channels.length === 0) {
```
Comment (virdesai) [inline src/components/ChannelList/ChannelListMessenger.js:118]:
```suggestion
  if (error && !refreshing && !channels.length) {
```

**Pull request merged**